### ISSUE TITLE
List active data provider question

### ISSUE THREAD

**Issue by layanto** [deepstreamIO/deepstream.io-client-js#306]
Is this how I should code active data provider for list? I am following the suggestion in https://github.com/deepstreamIO/deepstream.io-client-js/issues/276 to only accept the request in the listener when data is available.

```
        ds.record.listen('list/last-5-comments', (match, isSubscribed, response) => {
            if(isSubscribed) {
                console.log('subscribed')
                const list = ds.record.getList(match)
                //Periodically query db (maybe via RethinkDb's changefeeds) and populate/update list
                list.setEntries(['123', '456'])
                response.accept()
            } else {
                // stop publishing data
                console.log('unsubscribed')
                ds.record.getList(match).discard()
            }
        })
```

Do I need to keep track of the number of subscribers, how do I do that? How do I find out when all subscribers have un-subscribed? When all subscribers have un-subscribed, how do I delete the list? I tried replacing discard() with delete() and this ended up crashing deepstream server.

**Comment by layanto** [comment]
I did some trial and error today and made the following observations.

With listener logic that does not provide data (ie not active data provider)
```
ds.record.listen('list/last-5-comments', (match, isSubscribed, response) => {
            if(isSubscribed) {
                console.log('subscribed')
                response.accept()
            } else {
                console.log('unsubscribed')
            }
        })
```
console.log('subscribed') is called only once, when the first client subscribes and console.log('unsubscribed') is called only once, when the last client unsubscribes. This matches the description provided in https://deepstream.io/tutorials/core/datasync-records/.
However, it is different for active data provider as the listener is also a subscriber. The following comment is in deepstream.io-provider-search-rethinkdb https://github.com/deepstreamIO/deepstream.io-provider-search-rethinkdb/blob/master/src/search.js
```
 * Important: Deepstream notifies this provider once all subscriptions for list have been
 * removed - which presents a conundrum since the provider is a subscriber in its own right,
 * so will never be notified. Instead it listens to the list being deleted - which isn't an ideal
 * solution since it either requires namespacing the lists with a username or deleting the lists
 * for all subscribers.
```
If this is true, how can active data provider ever work? Active data provider will start providing data when the first client subscribes, but because the active data provider itself is also a subscriber, it doesn't know when the last client unsubscribes, as such continue to provide data when no-one is listening for the data. Is this the case??
Deepstream server should not consider a listener client as a subscriber just as a deepstream client publishing a message via pub/sub does not receive its own message.

But with the following listener logic
```
    var list
    ds.record.listen('list/last-5-comments', (match, isSubscribed, response) => {
            if(isSubscribed) {
                console.log('subscribed')
                list = ds.record.getList(match)
                list.setEntries(['123', '456'])
                response.accept()
            } else {
                // stop publishing data
                console.log('unsubscribed')
                list.discard()
            }
        })
```
When the last client unsubscribes, console.log('unsubscribed') is called, followed by console.log('subscribed') and then followed by 2x unsolicited_message R, ack_timeout R and response_timeout R. Why is that? Based on the comment in rethinkdb search provider, console.log('unsubscribed') should never be called as even after the last client unsubscribes, the listener itself is a subscriber.

In summary, I am thoroughly confused now...

**Comment by layanto** [comment]
Did more reading, I believe active data provider should work from deepstream server 1.1.0 onwards, meaning the comment in provider-search-rethinkdb is probably obsolete. But I still can't get this to work.

I tried the same logic using deepstream server 1.1.2 and client-js 1.1.0 and it mostly behaved as I expected. No console.log('subscribed) following console.log('unsubscribed') when last client unsubscribes. Although there were a couple of UNSOLICITED_MESSAGE R. I can even do list.delete() to delete the list when last client unsubscribes. Doing the same on latest deepstream server and client-js resulted in repeated subscribed and unsubscribed and deepstream server crashing.

Is active data provider broken in deepstream server 2.1.2 and client-js 2.1.1?

**Comment by WolframHempel** [comment]
Hi @layanto , sorry for the late reply.

Fundamentally, this problem has been addressed with an extension to the data-provider API that introduced `response.accept()` in deepstream 1.1.0. Here's what happens:

- multiple processes can register as listeners for a pattern
- upon subscription deepstream will try them **sequentially** in a random order to distribute load
- as you said, this doesn't happen on every subscription but only for the **first** time a client subscribes and on the last time a client unsubscribes
- the first listener to call `response.accept()` will be considered the provider
- other listeners have to call `response.reject()` or timeout
- **the listener that calls `response.accept()` is **NOT** counted towards the subscribers

So basically 
- make sure to call `response.accept()` - otherwise your listener will just be another subscriber
- there shouldn't be a need to delete lists to trigger unsubscribe. The reason the RehtinkDB search provider does it is that the lists are temporary

Please let me know if that solves your problem

**Comment by layanto** [comment]
With deepstream server 2.1.2 and client-js 2.1.1
Backend client (active data provider):
```
const deepstream = require('deepstream.io-client-js')
const ds = deepstream('127.0.0.1:6020')

var client = ds.login({username: 'backend'})
client.record.listen('list/testing', (match, isSubscribed, response) => {
    if(isSubscribed) {
        console.log('subscribed ' + match)
        client.record.getList(match).setEntries(['123', '456'])
        response.accept()
    } else {
        console.log('unsubscribed ' + match)
        client.record.getList(match).discard()
        //client.record.getList(match).delete()  this crashes deepstream server
    }
})

client.on('error', (error, event, topic) => {
    console.log(error, event, topic)
})
```
Subscriber client:
```
const deepstream = require('deepstream.io-client-js')
const ds = deepstream('127.0.0.1:6020')

var client = ds.login({username: 'client'})
list = client.record.getList('list/testing').whenReady((list) => {
    list.subscribe((entries) => {
        console.log(entries)
    })
})

client.on('error', (error, event, topic) => {
    console.log(error, event, topic)
})
```
From console of backend client:
```
subscribed list/testing
list/testing|0 UNSOLICITED_MESSAGE R ------ why is there unsolicited message?
list/testing|T UNSOLICITED_MESSAGE R
```
When I close the subscribe client, from console of backend client:
```
unsubscribed list/testing
subscribed list/testing  ------ why is the listener re-subscribed again?
list/testing|F UNSOLICITED_MESSAGE R ------ why is there unsolicited message?
list/testing|T UNSOLICITED_MESSAGE R
```

**Comment by layanto** [comment]
With deepstream server 1.1.2 and client-js 1.1.0
Backend client (active data provider):
```
const deepstream = require('deepstream.io-client-js')
const ds = deepstream('127.0.0.1:6021')

var client = ds.login({username: 'backend'})
client.record.listen('list/testing', (match, isSubscribed, response) => {
    if(isSubscribed) {
        console.log('subscribed ' + match)
        client.record.getList(match).setEntries(['123', '456'])
        response.accept()
    } else {
        console.log('unsubscribed ' + match)
        client.record.getList(match).discard()
        //client.record.getList(match).delete()  this works too
    }
})

client.on('error', (error, event, topic) => {
    console.log(error, event, topic)
})
```
Subscriber client:
```
const deepstream = require('deepstream.io-client-js')
const ds = deepstream('127.0.0.1:6021')

var client = ds.login({username: 'client'})
list = client.record.getList('list/testing').whenReady((list) => {
    list.subscribe((entries) => {
        console.log(entries)
    })
})

client.on('error', (error, event, topic) => {
    console.log(error, event, topic)
})
```
From console of backend client:
```
subscribed list/testing
list/testing|0 UNSOLICITED_MESSAGE R ------ why is there unsolicited message?
list/testing|T UNSOLICITED_MESSAGE R
```
When I close the subscribe client, from console of backend client:
```
unsubscribed list/testing
list/testing|F UNSOLICITED_MESSAGE R ------ why is there unsolicited message?
```

**Comment by layanto** [comment]
Note that with deepstream server 1.1.2 and client-js 1.1.0, when client unsubscribes, listener is not re-subscribed as was the case with deepstream server 2.1.2 and client-js 2.1.1

**Comment by WolframHempel** [comment]
could you change your backend provider as follows

```javascript
const deepstream = require('deepstream.io-client-js')
const ds = deepstream('127.0.0.1:6020')

var client = ds.login({username: 'backend'})
client.record.listen('list/testing', (match, isSubscribed, response) => {
    var lists = {};

    if(isSubscribed) {
        // accept first
        response.accept()
        console.log('subscribed ' + match)
        lists[ match ] = client.record.getList( match )
        lists[ match ].setEntries(['123', '456'])
        // here you could set more / other entries as they become available
        
    } else {
        console.log('unsubscribed ' + match)

        // calling getList() creates a nother subscription 
        // on this client and is the opposite operation of discard
        // instead discard the original list
        lists[ match ].discard()
    }
})

client.on('error', (error, event, topic) => {
    console.log(error, event, topic)
})
```

UNSOLICITED_MESSAGE is more of an info than an error. It occurs because the provider receives data for the list after it already called `discard()`. You can either ignore it or change the provider to

```javascript
client.record.getList( match ).whenReady( list => {
    list.setEntries(['123', '456'])
    list.discard()
})
```

**Comment by layanto** [comment]
Now getting lists[ match ].discard() TypeError: Cannot read property 'discard' of undefined after unsubscribed list/testing

**Comment by layanto** [comment]
The reason I do response.accept() after setEntries is because that was the proposed work around to make sure that the list is populated with data by checking hasProvider - see https://github.com/deepstreamIO/deepstream.io-client-js/issues/276

**Comment by layanto** [comment]
Need to move var lists = {} outside of client.record.listen

**Comment by layanto** [comment]
But I still get subscribed list/testing after unsubscribed list/testing, meaning the active data provider will still be active

**Comment by layanto** [comment]
On client subscribes, console of backend provider shows:
```
subscribed list/testing
list/testing|T UNSOLICITED_MESSAGE R
list/testing|l UNSOLICITED_MESSAGE R
```
On client unsubscribes:
```
unsubscribed list/testing
subscribed list/testing
list/testing|F UNSOLICITED_MESSAGE R
list/testing|T UNSOLICITED_MESSAGE R
list/testing|l UNSOLICITED_MESSAGE R
list/testing ACK_TIMEOUT R
list/testing RESPONSE_TIMEOUT R
```

**Comment by layanto** [comment]
If I do response.accept() after setEntries, sometimes deepstream server crash due to infinite cycle of subscribed and unsubscribed when client unsubscribes.

**Comment by WolframHempel** [comment]
oh dear, ... ok, one after the other: 
- you're right - `var lists` needs to be outside of the callback
- the server crash is worrying - what exactly happens? Any console output?
- if a subscription that matches is removed, your provider will be notified, irrespective if its already been notified of the subscription being added. So yes, we would need to add a `if( lists[match] )` check

**Comment by layanto** [comment]
1. On closer look, deepstream server only crashes on my original code where I call client.record.getList(match).delete() when last client unsubscribes. The crash can happen whether response.accept() is called before or after client.record.getList(match).setEntries(['123', '456']). I understand now that I should not call getList again and should call discard of the original list but this IMHO shouldn't crash deepstream server. Otherwise, any deepstream client with listen permission can easily crash the server. The crash happened (I believe) due to infinite cycle of subscribe and unsubscribe and delete the list; from memory the crash was relating to undefined type logger(?). On testing again this morning, deepstream server wasn't crashing but was keeping CPU usage high indefinitely (until I stop the backend client listener/provider). Possible DoS?
2. Why is the listener getting called again and triggering console.log('subscribed ' + match) after the last client unsubscribes? This doesn't happen with deepstream server 1.1.2
3. Question 2 above may be related to your third bullet above which unfortunately I don't fully understand (where do I need to put the if( lists[match] ) check?). I was under the impression that the listener call back is only called (with isSubscribed = TRUE) when the first client subscribes and again (with isSubscribed = FALSE) when the last client unsubscribes.

**Comment by layanto** [comment]
Did more testing and deepstream server 2.0.1 with client-js 2.0.0 behaves correctly as are deepstream server 1.1.2 and client-js 1.1.0.
I changed the code to the following
Backend client/provider:
```javascript
const deepstream = require('deepstream.io-client-js')
const ds = deepstream('127.0.0.1:6021')

var client = ds.login({username: 'backend'})
var lists = {}
var intervals = {}

client.record.listen('list/testing', (match, isSubscribed, response) => {

    if(isSubscribed && typeof lists[ match ] === 'undefined') {
        console.log('subscribed ' + match)
        lists[ match ] = client.record.getList( match )
        lists[ match ].whenReady((list) => {
            response.accept()
            var count = 0
            intervals[ match ] = setInterval(() => {
                console.log('ping')
                count = count + 1
                list.addEntry(count.toString())
            }, 1000)
        })
    } else {
            console.log('unsubscribed ' + match)
            console.log('deleting ' + match)
            lists[ match ].delete()
            delete lists[ match ]
            clearInterval( intervals[ match ] )
            delete intervals[ match ]
        }
    }
})

client.on('error', (error, event, topic) => {
    console.log(error, event, topic)
})
``` 
Client:
```javascript
const deepstream = require('deepstream.io-client-js')
const ds = deepstream('127.0.0.1:6020')

var client = ds.login({username: 'client'})
list = client.record.getList('list/testing').whenReady((list) => {
    list.subscribe((entries) => {
        console.log(entries)
    }, true)
})

client.on('error', (error, event, topic) => {
    console.log(error, event, topic)
})
```

**Comment by layanto** [comment]
Tested deepstream server 2.1.0 and client-js 2.1.0 and not working correctly. Something between 2.0.1 and 2.1.0 broke this feature.

**Comment by layanto** [comment]
Video (gif) of correct behaviour on deepstream server 2.0.1
![ds201](https://cloud.githubusercontent.com/assets/245078/21713951/a2939234-d450-11e6-9b92-ca8dcc4fcb34.gif)

Video (gif) of incorrect behaviour on deepstream server 2.1.0
![ds210](https://cloud.githubusercontent.com/assets/245078/21714011/fd21fcea-d450-11e6-84eb-e9667c6da57e.gif)

The was a couple of times where somehow the backend client/provider still generating 'ping' on console even when client has disconnected (possibly due to the subscribe callback being triggered immediately after unsubscribe).

**Comment by layanto** [comment]
Deepstream server crash on logger error is the issue raised in https://github.com/deepstreamIO/deepstream.io/issues/527

**Comment by layanto** [comment]
@WolframHempel Sorry chasing you up for updates. Is this an issue with deepstream 2.1.x or is there problem with my implementation? With same code it works as expected with deepstream 2.0.1 but not with 2.1.x. I am stuck at the moment with the only option reverting to deepstream 2.0.1.

**Comment by WolframHempel** [comment]
sorry about that - generally deepstream should be able to receive any kind of malformed data or update without crashing. Thanks for providing the detailed description, I'll keep you posted in this thread

**Comment by jdmnd** [comment]
Hi @layanto.
Thanks for providing such a detailed report! I've just opened an issue on the main deepstream repository that I believe relates to your issue:
https://github.com/deepstreamIO/deepstream.io/issues/531

Hopefully we'll have a release out later today with a fix. 

I haven't been able to replicate a server crash, however.

**Comment by yasserf** [comment]
Small change in plans, we are planning on getting the release out on Friday!

**Comment by layanto** [comment]
Regarding server crash, with deepstream server 2.1.2 configured to use redis for cache and message and rethinkdb for storage, and with incorrect listener code below. Note that I am now aware that this listener logic is incorrect as during unsubscribe, the logic to delete list is in fact re-subscribing causing repeating cycle of unsubscribe and subscribe.
```javascript
const deepstream = require('deepstream.io-client-js')
const ds = deepstream('127.0.0.1:6020')

var client = ds.login({username: 'backend'})
client.record.listen('list/testing', (match, isSubscribed, response) => {
    if(isSubscribed) {
        console.log('subscribed ' + match)
        client.record.getList(match).setEntries(['123', '456'])
        response.accept()
    } else {
        console.log('unsubscribed ' + match)
        client.record.getList(match).delete()
    }
})

client.on('error', (error, event, topic) => {
    console.log(error, event, topic)
})
```
Client code:
```javascript
const deepstream = require('deepstream.io-client-js')
const ds = deepstream('127.0.0.1:6020')

var client = ds.login({username: 'client'})
list = client.record.getList('list/testing')
list.subscribe((entries) => {
    console.log(entries)
})

client.on('error', (error, event, topic) => {
    console.log(error, event, topic)
})
```
See video (gif) below:
![ds212-infinite loop](https://cloud.githubusercontent.com/assets/245078/21885715/00188188-d90d-11e6-84c5-56603d81165b.gif)

If deepstream server 2.1.2 is not configured to use redis for cache/message and rethinkdb for storage, the same logic will result in infinite unsubscribe and subscribe but the server does not crash, effectively a DoS. See video (gif) below:
![ds212-crash](https://cloud.githubusercontent.com/assets/245078/21885751/2fc332f2-d90d-11e6-922a-a622f616f289.gif)

**Comment by layanto** [comment]
Appears to be fixed in ds 2.1.3 and js-client 2.1.2
